server/lib/translator.py

```diff
@@ -21,42 +22,168 @@
+def translate_page_config(page_config: dict, query_lang: str,
+                          counters: Counters) -> dict:
+  try:
+    original_strings = _get_strings_for_translation(page_config)
+    strings_with_tokens: list[TranslationStringTokens] = []
+    strings_for_translation: list[str] = []
+    for original_string in original_strings:
+      translation_string_tokens = TranslationStringTokens(original_string)
+      strings_with_tokens.append(translation_string_tokens)
+      strings_for_translation.append(
+          translation_string_tokens.string_for_translation)
+
+    translated_strings = _translate(queries=strings_for_translation,
+                                    source_lang=_EN_LANG,
+                                    target_lang=query_lang)
+
+    translated_strings_with_tokens_reinserted = []
+    for translated_string, translation_string_token in zip(
+        translated_strings, strings_with_tokens):
+      translated_strings_with_tokens_reinserted.append(
+          translation_string_token.reinsert_tokens(translated_string))
+    translations = dict(
+        zip(original_strings, translated_strings_with_tokens_reinserted))
+
+    counters.info(
+        "translate_page_config",
+        {
+            "num_translations": len(strings_with_tokens),
+            "translations": translations
+        },
+    )
+
+    _populate_translated_strings(page_config, translations)
+
+    return page_config
+  except Exception as e:
+    counters.err("translate_page_config_error", {"error": str(e)})
+    return page_config
+
+
+def _populate_translated_strings(page_config: dict, translations: dict[str,
+                                                                       str]):
+
+  def _maybe_update(obj: dict, field: str):
+    original = obj.get(field, "")
+    if original and original in translations:
+      obj.update({field: translations[original]})
+
+  for category in page_config.get("categories", []):
+    for block in category.get("blocks", []):
+      _maybe_update(block, "title")
+      for column in block.get("columns", []):
+        for tile in column.get("tiles", []):
+          _maybe_update(tile, "description")
+          _maybe_update(tile, "title")
+    for sv in category.get("statVarSpec", {}).values():
+      _maybe_update(sv, "name")
+
+
+def _get_strings_for_translation(
+    page_config: dict) -> list[TranslationStringTokens]:
+  strings = set()
+
+  def _maybe_add(string: str):
+    if string:
+      strings.add(string)
+
+  for category in page_config.get("categories", []):
+    for block in category.get("blocks", []):
+      _maybe_add(block.get("title"))
+      for column in block.get("columns", []):
+        for tile in column.get("tiles", []):
+          _maybe_add(tile.get("description"))
+          _maybe_add(tile.get("title"))
+    for sv in category.get("statVarSpec", {}).values():
+      _maybe_add(sv.get("name"))
+
+  return list(strings)
```

Comment:
I wonder if the logic would be simpler (without the need for `TranslationStringTokens`), if we just stripped the `${date}`-like stuff in the returned strings (in _maybe_add), and again when checking the map (in _maybe_update)?